packages/pds/src/app-migrations/update-follow-subjects.ts

```diff
@@ -0,0 +1,169 @@
+import { cborBytesToRecord, chunkArray } from '@atproto/common'
+import { sql } from 'kysely'
+import AppContext from '../context'
+import Database from '../db'
+import { appMigration } from '../db/leader'
+import { MessageDispatcher } from '../event-stream/message-queue'
+import { ids } from '../lexicon/lexicons'
+import { prepareUpdate, prepareDelete, assertValidRecord } from '../repo'
+import { RepoService } from '../services/repo'
+
+const MIGRATION_NAME = '2023-03-14-update-follow-subjects'
+const SHORT_NAME = 'update-follow-subjects'
+
+export async function updateFollowSubjectsMigration(ctx: AppContext) {
+  await appMigration(ctx.db, MIGRATION_NAME, (tx) => main(tx, ctx))
+}
+
+async function main(tx: Database, ctx: AppContext) {
+  console.log(SHORT_NAME, 'beginning')
+  tx.assertTransaction()
+  const { ref } = tx.db.dynamic
+  const now = new Date().toISOString()
+
+  // The message dispatcher usually ensures that the app view indexes these updates,
+  // but that has been taken care of via a db migration to remove the indexes entirely.
+  const noopDispatcher = new MessageDispatcher()
+  noopDispatcher.destroy()
+
+  const repoTx = new RepoService(
+    tx,
+    ctx.repoSigningKey,
+    noopDispatcher,
+    ctx.blobstore,
+  )
+
+  // For each user update follow records.
+
+  const [allDids, followCount] = await Promise.all([
+    getAllDids(tx),
+    getFollowCount(tx),
+  ])
+
+  console.log(
+    SHORT_NAME,
+    `${followCount} updates across ${allDids.length} dids`,
+  )
+
+  let didsComplete = 0
+  let updatesComplete = 0
+  const updatesTurnedDeletes: string[] = []
+  const chunks = chunkArray(allDids, Math.ceil(allDids.length / 10))
+
+  await Promise.all(
+    chunks.map(async (dids) => {
+      for (const did of dids) {
+        const follows = await getFollows(tx, did)
+        console.log(SHORT_NAME, `${did} processing ${follows.length} updates`)
+
+        const updates = await Promise.all(
+          follows.map((follow) => {
+            const record = cborBytesToRecord(follow.bytes)
+            if (typeof record['subject']?.['did'] === 'string') {
+              // Map old app.bsky.actor.ref to did
+              record['subject'] = record['subject']['did']
+            }
+            try {
+              assertValidRecord(record)
+              return prepareUpdate({
+                did,
+                collection: ids.AppBskyGraphFollow,
+                rkey: follow.rkey,
+                record,
+                validate: false,
+              })
+            } catch {
+              const del = prepareDelete({
```

Comment:
just curious, how many hit this in trial runs?

Reply:
There were 7, I think. I looked at them all and they were indeed malformed in some way (either a handle rather than a did, or empty did).